File: src/expr.cpp
#include "expr.hpp"

int Expr::count = 0;

#include "garbage_collector.hpp"
#include "literal.hpp"

Expr::Expr() {
  count++;
  GarbageCollector<Expr>::Push(this);
}

Expr::Expr(int ln) {
  line = ln;
  count++;
  GarbageCollector<Expr>::Push(this);
}

Expr::~Expr() {
  count--;
  GarbageCollector<Expr>::Pop(this);
}

void Expr::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}
Array::Array(int ln, std::vector<Expr*> exprVector) {
    this->line = ln;
    this->exprVector = exprVector;
  }

Array::~Array() {
}

void Array::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

Assign::Assign(int ln, Expr* target, Expr* value) {
    this->line = ln;
    this->target = target;
    this->value = value;
  }

Assign::~Assign() {
}

void Assign::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

Binary::Binary(int ln, Expr* lhs, Token op, Expr* rhs) {
    this->line = ln;
    this->lhs = lhs;
    this->op = op;
    this->rhs = rhs;
  }

Binary::~Binary() {
}

void Binary::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

Class::Class(int ln, Block* block) {
    this->line = ln;
    this->block = block;
  }

Class::~Class() {
}

void Class::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

Function::Function(int ln, std::vector<std::string> parameterVector, Block* block) {
    this->line = ln;
    this->parameterVector = parameterVector;
    this->block = block;
  }

Function::~Function() {
}

void Function::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

Grouping::Grouping(int ln, Expr* inner) {
    this->line = ln;
    this->inner = inner;
  }

Grouping::~Grouping() {
}

void Grouping::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

Index::Index(int ln, Expr* array, Expr* index) {
    this->line = ln;
    this->array = array;
    this->index = index;
  }

Index::~Index() {
}

void Index::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

Invocation::Invocation(int ln, Expr* expr, std::vector<Expr*> exprVector) {
    this->line = ln;
    this->expr = expr;
    this->exprVector = exprVector;
  }

Invocation::~Invocation() {
}

void Invocation::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

Logical::Logical(int ln, Expr* lhs, Token op, Expr* rhs) {
    this->line = ln;
    this->lhs = lhs;
    this->op = op;
    this->rhs = rhs;
  }

Logical::~Logical() {
}

void Logical::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

Unary::Unary(int ln, Token op, Expr* rhs) {
    this->line = ln;
    this->op = op;
    this->rhs = rhs;
  }

Unary::~Unary() {
}

void Unary::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

Value::Value(int ln, Literal* value) {
    this->line = ln;
    this->value = value;
  }

Value::~Value() {
}

void Value::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

Variable::Variable(int ln, Token name) {
    this->line = ln;
    this->name = name;
  }

Variable::~Variable() {
}

void Variable::Accept(ExprVisitor* visitor) {
  visitor->Visit(this);
}

File: src/Expr.java
package com.krgamestudios.toy;

import java.util.List;

abstract class Expr {
	interface Visitor<R> {
		R visit(Assign expr);
		R visit(Binary expr);
		R visit(Grouping expr);
		R visit(Literal expr);
		R visit(Logical expr);
		R visit(Postfix expr);
		R visit(Prefix expr);
		R visit(Variable expr);
	}

	abstract <R> R accept(Visitor<R> visitor);
	static class Assign extends Expr {
		final Token name;
		final Expr value;

		Assign(Token name, Expr value) {
			this.name = name;
			this.value = value;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class Binary extends Expr {
		final Expr left;
		final Token operator;
		final Expr right;

		Binary(Expr left, Token operator, Expr right) {
			this.left = left;
			this.operator = operator;
			this.right = right;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class Grouping extends Expr {
		final Expr expression;

		Grouping(Expr expression) {
			this.expression = expression;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class Literal extends Expr {
		final Object value;

		Literal(Object value) {
			this.value = value;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class Logical extends Expr {
		final Expr left;
		final Token operator;
		final Expr right;

		Logical(Expr left, Token operator, Expr right) {
			this.left = left;
			this.operator = operator;
			this.right = right;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class Postfix extends Expr {
		final Expr left;
		final Token operator;

		Postfix(Expr left, Token operator) {
			this.left = left;
			this.operator = operator;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class Prefix extends Expr {
		final Token operator;
		final Expr right;

		Prefix(Token operator, Expr right) {
			this.operator = operator;
			this.right = right;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class Variable extends Expr {
		final Token name;

		Variable(Token name) {
			this.name = name;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
}
